### PR TITLE
Feature: Option to skip hooks

### PR DESCRIPTION
**Pull request by noirbizarre**
## Pull Request Check List

- [x] A news fragment is added in `news/` describing what is new.
- [x] Test cases added for changed code.

## Describe what you have changed in this PR.

This PR is a proposal at implementing #948.
If provides 2 parameters:
- `--skip-hooks` that will skip any hook (or pre/post script) involved in the execution
- `--skip` which can take any arbitrary task or hook name(s) and skip it/them

Those 2 parameters have been added to the following commands
- `run`
- `install`
- `lock`
- `build`
- `init`

The `--skip` can be used many times. The following commands are all equivalents
- `--skip=pre_task1,task2`
- `--skip pre_task1,task2`
- `--skip=pre_task1 --skip=task2`
- `--skip pre_task1 --skip task2`

The both options can be used together:
```console
pdm run --skip-hooks --skip=taskn my-composite-task
```

### CODE REVIEW

**Comment by frostming** [comment]
> Those 2 parameters have been added to the following commands
> 
> * `run`
> * `install`
> * `lock`
> * `build`
> * `init`

That is not complete, the following commands also call hooks:

- `add`
- `update`
- `remove`
- `publish` (new in 2.0)

Besides, can we merge the two options `--skip-hooks` and `--skip` into one? And ideally, the hooks should also be able to control with env vars.

**Comment by noirbizarre** [comment]
I'll add the option to those extra commands and I think documentation needs an update on this (they are not listen) or at least an explanation that signals  does not match the commands but the actions executed by commands (if I understand well, install can trigger `pre_lock` and `post_lock`...). I'll try addind some lifecycle documentation somewhere (or update it if it already exists)

For the `:all` keyword, I'll do as well as `:pre` and `:post`  with:
- `:all`: skip all hooks
- `:pre`: skip all `pre` hooks
- `:post`: skip all `post` hooks

**Comment by frostming** [comment]
> I'll try adding some lifecycle documentation somewhere (or update it if it already exists)

A Lifecycle documentation would be superb!

> For the :all keyword, I'll do as well as :pre and :post with:
> 
> :all: skip all hooks
> :pre: skip all pre hooks
> :post: skip all post hooks

Yeah, I love the shortcuts

File: pdm/cli/commands/run.py [hunk 1]
```diff
@@ -223,25 +228,38 @@ def _run_task(
...
         if task is not None:
+
             pre_task = self._get_task(f"pre_{command}")
-            if pre_task is not None:
+            if (
+                pre_task is not None
+                and not self.hooks.skip_all
+                and not self.hooks.skip_pre
+                and pre_task.name not in self.hooks.skip
+            ):
```
Comment (frostming) [inline pdm/cli/commands/run.py:244]: I love the abstraction of `HookManager` but these lines seem to be repeating. What about adding a method `HookManager.should_skip(name: str) -> bool` to check for that?

File: pdm/cli/hooks.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    def skip_post(self) -> bool:
+        return ":post" in self.skip
+
+    def try_emit(self, name: str, **kwargs: Any) -> None:
+        if (
+            not self.skip_all
+            and name not in self.skip
+            and not (self.skip_pre and name.startswith("pre_"))
+            and not (self.skip_post and name.startswith("post_"))
+        ):
```
Comment (frostming) [inline pdm/cli/hooks.py:45]: These can also benefit from the `should_skip(name)` method.

File: pdm/cli/options.py [hunk 3]
```diff
@@ -230,6 +250,14 @@ def no_isolation_callback(
     help="Save minimum version specifiers",
 )
 
+skip_option = Option(
+    "-k",
+    "--skip",
+    dest="skip",
+    action=split_lists(","),
+    help="Skip some tasks and/or hooks",
```
Comment (frostming) [inline pdm/cli/options.py:258]: Read from env var `PDM_SKIP_HOOKS` if the option is not given?

**Review by frostming — CHANGES_REQUESTED** [review]
@noirbizarre Thanks for your work on that feature. The code is of high quality and your Python skills are awesome. I've made some comments and improvement suggestions. Also, would you like to write a new documentation page about hooks?

File: news/1127.feature.md [hunk 4]
```diff
@@ -0,0 +1 @@
+Add 2 new executions options: `--skip-hooks` and `--skip` to opt-out some scripts and hooks from any execution,
```
Comment (frostming) [inline news/1127.feature.md:1]: This line needs to be updated.

File: tests/cli/test_hooks.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+@parametrize_with_commands
+def test_hooks(hooked_project, invoke, capfd, specs: HookSpecs):
+    invoke(shlex.split(specs.command), strict=True, obj=hooked_project)
+    out, _ = capfd.readouterr()
+    for hook in specs.hooks:
+        assert f"{hook} CALLED" in out
+
+
+@parametrize_with_hooks  # Iterate over hooks as wee need a clean slate for each run
```
Comment (Hnasar) [inline tests/cli/test_hooks.py:197]: Typo

**Comment by Hnasar** [comment]
@frostming I've been meaning to add a bug for a while that it seems like the relevant hooks aren't called from pdm-pep517 (e.g. if you add a pre_build hook it doesn't get called from python -m build).

If hooks are a fundamental part of PDM (and not just a part of the CLI) would it make sense to move HookManager into pdm-pep517 and then import it here?

**Comment by frostming** [comment]
> @frostming I've been meaning to add a bug for a while that it seems like the relevant hooks aren't called from pdm-pep517 (e.g. if you add a pre_build hook it doesn't get called from python -m build).
> 
> If hooks are a fundamental part of PDM (and not just a part of the CLI) would it make sense to move HookManager into pdm-pep517 and then import it here?

I have a different opinion on that. Lifecycle scripts should be considered part of the package manager PDM, i.e. the **frontend**, rather than pdm-pep517, the **backend**. Like in node.js world, Yarn and npm support different lifecycle scripts. If you use other frontends, they may not call the hooks.

File: pdm/cli/commands/run.py (hunk 1 above)
Comment (noirbizarre) [reply pdm/cli/commands/run.py]: I've gone for `HookManager.should_run(name: str) -> bool` as it avoids double negations.
It is more easy to read.

File: pdm/cli/hooks.py (hunk 2 above)
Comment (noirbizarre) [reply pdm/cli/hooks.py]: Indeed, yes it does 👌🏼

File: pdm/cli/options.py (hunk 3 above)
Comment (noirbizarre) [reply pdm/cli/options.py]: Done 👍🏼 
It's not easily testable as `Option.default` is set once on parsing and can't change during tests but given I have to process the splitted list, I tested the method used to generate the `default`.

**Comment by noirbizarre** [comment]
> @noirbizarre Thanks for your work on that feature. The code is of high quality and your Python skills are awesome. I've made some comments and improvement suggestions. Also, would you like to write a new documentation page about hooks?

👆🏼 Thanks 🙏🏼 🙏🏼 

I applied changes/suggestions and rebased on the latest `dev` to resolve conflicts.
Do you want the hooks documentation in this PR or in a separate one (because it might take some times to write a good one and holding the PR might lead to more conflict solving).

Another question: do you me to add a `pre_publish` and `post_publish` hook ? Seems to be some legit and useful hooks

**Comment by frostming** [comment]
> I applied changes/suggestions and rebased on the latest dev to resolve conflicts and squashed some commits to make history more consistent.

Thanks, hope my updates to the dev branch don't cause too much trouble. The CI seems passing.

> Do you want the hooks documentation in this PR or in a separate one

A separate PR would be okay.

> do you me to add a pre_publish and post_publish hook ? Seems to be some legit and useful hooks

Sure, go ahead for it.

**Comment by noirbizarre** [comment]
> > I applied changes/suggestions and rebased on the latest dev to resolve conflicts and squashed some commits to make history more consistent.
> >
> Thanks, hope my updates to the dev branch don't cause too much trouble. The CI seems passing.
>
As soon as I understood `dev` has been rebased, no problems 👍🏼 

> > Do you want the hooks documentation in this PR or in a separate one
> 
> A separate PR would be okay.
> 
OK, let's go this way 👍🏼 

> > do you me to add a pre_publish and post_publish hook ? Seems to be some legit and useful hooks
> 
> Sure, go ahead for it.

Great, I'll add then in another PR too


So the PR should be good to merge and I'll submit extra changes (`pre_publish`/`post_publish` and extended hooks documentation) in separate pull-requests 👍🏼 

(And thanks for the reactivity, it's a pleasure contributing with quick feedback)

**Review by frostming — APPROVED** [review]
LGTM

**Comment by noirbizarre** [comment]
Great 🎊 
I'll let you merge, I don't have the proper permissions to do so.

**Pull request merged**